internal/pkg/policy/secret.go

```diff
@@ -0,0 +1,111 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package policy
+
+import (
+	"context"
+	"encoding/json"
+	"regexp"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+)
+
+type SecretReference struct {
+	ID string `json:"id"`
+}
+
+// read secret values that belong to the agent policy's secret references, returns secrets as id:value map
+func getSecretReferences(ctx context.Context, secretRefsRaw json.RawMessage, bulker bulk.Bulk) (map[string]string, error) {
+	if secretRefsRaw == nil {
+		return nil, nil
+	}
+	var secretReferences []SecretReference
+	err := json.Unmarshal([]byte(secretRefsRaw), &secretReferences)
+	if err != nil {
+		return nil, err
+	}
+	ids := make([]string, 0)
+	for _, ref := range secretReferences {
+		ids = append(ids, ref.ID)
+	}
+	results, err := bulker.ReadSecrets(ctx, ids)
+	if err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+// read inputs and secret_references from agent policy
+// replace values of secret refs in inputs and input streams properties
+func getPolicyInputsWithSecrets(ctx context.Context, fields map[string]json.RawMessage, bulker bulk.Bulk) ([]map[string]interface{}, error) {
+	if fields["inputs"] == nil {
+		return nil, nil
+	}
+	var inputs []map[string]interface{}
+	err := json.Unmarshal([]byte(fields["inputs"]), &inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	if fields["secret_references"] == nil {
+		return inputs, nil
+	}
+
+	secretReferences, err := getSecretReferences(ctx, fields["secret_references"], bulker)
```

Comment:
naming nit: I think this function and it's return value should be called `getSecretValues` rather than references.